File: frontend/lib/models/public_library.dart
class PublicLibraryItem {
  const PublicLibraryItem({
    required this.id,
    required this.title,
    required this.author,
    required this.year,
    required this.kind,
    required this.genre,
    required this.level,
    required this.summary,
    required this.coverUrl,
    required this.textUrl,
    required this.externalUrl,
    required this.attribution,
    required this.sourceUrls,
    required this.license,
    required this.characters,
    required this.coverSourceUrl,
    required this.coverLicense,
    required this.coverAuthor,
  });

  final String id;
  final String title;
  final String author;
  final String year;
  final String kind;
  final String genre;
  final String level;
  final String summary;
  final String coverUrl;
  final String textUrl;
  final String externalUrl;
  final String attribution;
  final List<String> sourceUrls;
  final String license;
  final int characters;
  final String coverSourceUrl;
  final String coverLicense;
  final String coverAuthor;

  bool get isExternal => externalUrl.isNotEmpty && textUrl.isEmpty;

  factory PublicLibraryItem.fromJson(Map<String, dynamic> json) {
    return PublicLibraryItem(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      author: json['author'] as String? ?? '',
      year: json['year'] as String? ?? '',
      kind: json['kind'] as String? ?? '',
      genre: json['genre'] as String? ?? '',
      level: json['level'] as String? ?? '',
      summary: json['summary'] as String? ?? '',
      coverUrl: _absolute(json['coverUrl'] as String? ?? ''),
      textUrl: _absolute(json['textUrl'] as String? ?? ''),
      externalUrl: json['externalUrl'] as String? ?? '',
      attribution: json['attribution'] as String? ?? '',
      sourceUrls: (json['sourceUrls'] as List? ?? const [])
          .whereType<String>()
          .toList(growable: false),
      license: json['license'] as String? ?? '',
      characters: (json['characters'] as num?)?.toInt() ?? 0,
      coverSourceUrl: json['coverSourceUrl'] as String? ?? '',
      coverLicense: json['coverLicense'] as String? ?? '',
      coverAuthor: json['coverAuthor'] as String? ?? '',
    );
  }

  static String _absolute(String value) =>
      value.startsWith('/') ? 'https://citavuk.ru$value' : value;
}

File: web/src/lib/publicLibrary.ts
export interface PublicLibraryItem {
  id: string;
  title: string;
  author: string;
  year: string;
  kind: string;
  genre: string;
  level: string;
  summary: string;
  coverUrl: string;
  textUrl: string;
  externalUrl?: string;
  attribution?: string;
  sourceUrls: string[];
  license: string;
  characters: number;
  coverSourceUrl?: string;
  coverLicense?: string;
  coverAuthor?: string;
}

interface PublicLibraryCatalog {
  version: number;
  generatedAt: string;
  items: PublicLibraryItem[];
}

let catalogPromise: Promise<PublicLibraryCatalog> | null = null;

/** Загружает только компактные метаданные. Тексты запрашиваются по одному. */
export function loadPublicLibrary(
  signal?: AbortSignal,
): Promise<PublicLibraryCatalog> {
  if (!catalogPromise || signal) {
    const request = fetch('/public-library/catalog.json', {
      signal,
      cache: 'no-cache',
    }).then(async (response) => {
      if (!response.ok) throw new Error('Каталог временно недоступен.');
      return (await response.json()) as PublicLibraryCatalog;
    });
    if (!signal) catalogPromise = request;
    return request;
  }
  return catalogPromise;
}

export async function loadPublicBook(
  item: PublicLibraryItem,
  signal?: AbortSignal,
): Promise<string> {
  if (!item.textUrl) throw new Error('У этого материала нет локального текста.');
  const response = await fetch(item.textUrl, { signal });
  if (!response.ok) throw new Error('Не удалось загрузить текст.');
  return response.text();
}
